src/audio/module_adapter/module_adapter.c

```diff
@@ -623,54 +630,126 @@ static void module_adapter_process_output(struct comp_dev *dev)
 	}
 }
 
+static uint32_t
+module_single_sink_setup(struct comp_dev *dev,
+			 struct comp_buffer __sparse_cache **source_c,
+			 struct comp_buffer __sparse_cache **sinks_c)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct comp_copy_limits c;
+	struct list_item *blist;
+	uint32_t num_input_buffers = 0;
+	int i = 0;
+
+	list_for_item(blist, &dev->bsource_list) {
+		/* check if the source dev is in the same state as the dev */
+		if (!source_c[i]->source || source_c[i]->source->state != dev->state) {
+			i++;
+			continue;
+		}
+
+		comp_get_copy_limits_frame_aligned(source_c[i], sinks_c[0], &c);
+
+		buffer_stream_invalidate(source_c[i], c.frames * c.source_frame_bytes);
+
+		/*
+		 * note that the size is in number of frames not the number of
+		 * bytes
+		 */
+		mod->input_buffers[num_input_buffers].size = c.frames;
+		mod->input_buffers[num_input_buffers].consumed = 0;
+
+		mod->input_buffers[num_input_buffers].data = &source_c[i]->stream;
+		num_input_buffers++;
+		i++;
+	}
+
+	mod->output_buffers[0].size = 0;
+	mod->output_buffers[0].data = &sinks_c[0]->stream;
+
+	return num_input_buffers;
+}
+
+static uint32_t
+module_single_source_setup(struct comp_dev *dev,
+			   struct comp_buffer __sparse_cache **source_c,
+			   struct comp_buffer __sparse_cache **sinks_c)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct comp_copy_limits c;
+	struct list_item *blist;
+	uint32_t min_frames = UINT32_MAX;
+	uint32_t num_output_buffers = 0;
+	uint32_t source_frame_bytes = 0;
+	int i = 0;
+
+	list_for_item(blist, &dev->bsink_list) {
+		/* check if the sink dev is in the same state as the dev */
+		if (!sinks_c[i]->sink || sinks_c[i]->sink->state != dev->state) {
+			i++;
+			continue;
+		}
+
+		comp_get_copy_limits_frame_aligned(source_c[0], sinks_c[i], &c);
+
+		min_frames = MIN(min_frames, c.frames);
+		source_frame_bytes = c.source_frame_bytes;
+
+		mod->output_buffers[num_output_buffers].size = 0;
+		mod->output_buffers[num_output_buffers].data = &sinks_c[i]->stream;
+		num_output_buffers++;
+		i++;
+	}
+
+	buffer_stream_invalidate(source_c[0], min_frames * source_frame_bytes);
+	/* note that the size is in number of frames not the number of bytes */
+	mod->input_buffers[0].size = min_frames;
+	mod->input_buffers[0].consumed = 0;
+	mod->input_buffers[0].data = &source_c[0]->stream;
+
+	return num_output_buffers;
+}
+
 int module_adapter_copy(struct comp_dev *dev)
 {
 	struct processing_module *mod = comp_get_drvdata(dev);
 	struct module_data *md = &mod->priv;
 	struct comp_buffer *source, *sink;
 	struct comp_buffer __sparse_cache *source_c[PLATFORM_MAX_STREAMS];
+	struct comp_buffer __sparse_cache *sinks_c[PLATFORM_MAX_STREAMS];
 	struct comp_buffer __sparse_cache *sink_c = NULL;
-	struct comp_copy_limits c;
 	struct list_item *blist;
 	size_t size = MAX(mod->deep_buff_bytes, mod->period_bytes);
 	uint32_t min_free_frames = UINT_MAX;
 	uint32_t num_input_buffers = 0;
+	uint32_t num_output_buffers = 0;
```

Comment:
Not really specific to this PR, but the "copy()" starts to be very complex. I wonder if some of this logic could be moved to some less-often-run function. Could we even have a separate copy for the "simple_copy case"?